File: internal/server/prefs.go
package server

import (
	"context"
	"encoding/json"
	"net/http"
	"regexp"
	"strings"

	"github.com/rs/zerolog"

	"github.com/Zalaras/muster/internal/store"
)

// prefsKVKey is the single kv key prefs are persisted under, as one JSON blob
// (kb:anchor/prefs.put — no schema change).
const prefsKVKey = "prefs"

// defaultUsageModel is prefs.usageModel's default — the masthead's per-model readout
// before any PUT ever names one.
const defaultUsageModel = "Fable"

// defaultRailSort is prefs.railSort's default — the Focus rail's sort mode before any
// PUT ever names one.
const defaultRailSort = "manual"

// defaultTheme is prefs.theme's default — "follow" means the dashboard resolves its
// theme from claudeTheme.family; the daemon treats the value as opaque beyond
// validThemePattern.
const defaultTheme = "follow"

// defaultRailDensity is prefs.railDensity's default — the rail and Tiles-strip card
// density before any PUT ever names one (plan rail-card-improvements REQ-3).
const defaultRailDensity = "comfortable"

// defaultRailActivity is prefs.railActivity's default — which text a card's activity
// line shows before any PUT ever names one (plan rail-card-improvements REQ-13).
const defaultRailActivity = "turn"

// validThemePattern is prefs.theme's wire pattern (kb:anchor/prefs.put): 1-32 chars,
// a-z/0-9/-, starting with a letter. The daemon never interprets the value beyond this
// — the client owns the theme registry.
var validThemePattern = regexp.MustCompile(`^[a-z][a-z0-9-]{0,31}$`)

// prefsRequest is PUT /api/prefs' request body (kb:anchor/prefs.put): at least one
// field required, unknown fields ignored. Pointers distinguish "absent" from "present".
type prefsRequest struct {
	View         *string `json:"view"`
	Density      *string `json:"density"`
	UsageModel   *string `json:"usageModel"`
	RailSort     *string `json:"railSort"`
	Theme        *string `json:"theme"`
	UpdateCheck  *bool   `json:"updateCheck"`
	RailDensity  *string `json:"railDensity"`
	RailActivity *string `json:"railActivity"`
}

func validView(v string) bool     { return v == "focus" || v == "tiles" }
func validDensity(v string) bool  { return v == "2x2" || v == "3x2" }
func validRailSort(v string) bool { return v == "manual" || v == "attention" }
func validTheme(v string) bool    { return validThemePattern.MatchString(v) }

// validRailDensity/validRailActivity are prefs.railDensity/prefs.railActivity's enums
// (plan rail-card-improvements REQ-3/REQ-13).
func validRailDensity(v string) bool {
	return v == "compact" || v == "comfortable" || v == "expanded"
}
func validRailActivity(v string) bool {
	return v == "turn" || v == "prompt" || v == "reply" || v == "both"
}

// validUsageModel reports whether v (after trimming) is 1–32 chars (kb:anchor/prefs.put).
func validUsageModel(v string) bool {
	n := len(strings.TrimSpace(v))
	return n >= 1 && n <= 32
}

// PrefsInfo is the `prefs` object inside a snapshot (kb:anchor/prefs.put /
// kb:anchor/ws.snapshot), persisted in kv and broadcast on change. Density is the Tiles
// grid density. UsageModel is which modelScoped entry the masthead's third readout shows,
// default "Fable". RailSort is the rail's sort mode, "manual" | "attention", default
// "manual". Theme is opaque to the daemon beyond its pattern (kb:anchor/prefs.put),
// default "follow". UpdateCheck is whether the daemon checks GitHub Releases for a newer
// musterd, default true. RailDensity and RailActivity are the rail/Tiles-strip card
// density ("compact" | "comfortable" | "expanded", default "comfortable") and which text
// a card's activity line shows ("turn" | "prompt" | "reply" | "both", default "turn").
type PrefsInfo struct {
	View         string `json:"view"`
	Density      string `json:"density"`
	UsageModel   string `json:"usageModel"`
	RailSort     string `json:"railSort"`
	Theme        string `json:"theme"`
	UpdateCheck  bool   `json:"updateCheck"`
	RailDensity  string `json:"railDensity"`
	RailActivity string `json:"railActivity"`
}

// defaultPrefs is the shape before any PUT /api/prefs has ever landed (kb:anchor/prefs.put).
// UpdateCheck defaults true.
func defaultPrefs() PrefsInfo {
	return PrefsInfo{
		View: "focus", Density: "2x2", UsageModel: defaultUsageModel, RailSort: defaultRailSort,
		Theme: defaultTheme, UpdateCheck: true, RailDensity: defaultRailDensity, RailActivity: defaultRailActivity,
	}
}

// storedPrefs is loadPrefs' unmarshal target: UpdateCheck is a pointer here (unlike
// PrefsInfo's plain bool) so a persisted blob predating auto-update — which has no
// "updateCheck" key at all — is distinguishable from one that explicitly persisted
// false. A missing key defaults to true; an explicit false stays false; a non-boolean
// value fails the whole Unmarshal, which loadPrefs already treats as "return
// defaultPrefs()" (so it too loads as true).
type storedPrefs struct {
	View         string `json:"view"`
	Density      string `json:"density"`
	UsageModel   string `json:"usageModel"`
	RailSort     string `json:"railSort"`
	Theme        string `json:"theme"`
	UpdateCheck  *bool  `json:"updateCheck"`
	RailDensity  string `json:"railDensity"`
	RailActivity string `json:"railActivity"`
}

// prefsMessage is the WS `prefs` broadcast (kb:anchor/ws.prefs): a full-object echo
// of the persisted prefs, sent to every connected UI socket on every accepted PUT
// (INV-4).
type prefsMessage struct {
	Type  string    `json:"type"`
	Prefs PrefsInfo `json:"prefs"`
}

// loadPrefs reads the persisted prefs object from kv, falling back to defaults for a
// fresh daemon or an unreadable/corrupt value — a bad kv row must never fail a snapshot.
// A free function, not a prefsFeature method, since issueFeature's snapshot also needs
// it (dashboard scope's view/density/railSort row) without taking a dependency on the
// whole prefs feature.
func loadPrefs(ctx context.Context, st *store.Store) PrefsInfo {
	raw, ok, err := st.KVGet(ctx, prefsKVKey)
	if err != nil || !ok {
		return defaultPrefs()
	}
	var stored storedPrefs
	if err := json.Unmarshal([]byte(raw), &stored); err != nil {
		return defaultPrefs()
	}
	p := PrefsInfo{
		View:         stored.View,
		Density:      stored.Density,
		UsageModel:   stored.UsageModel,
		RailSort:     stored.RailSort,
		Theme:        stored.Theme,
		UpdateCheck:  true,
		RailDensity:  stored.RailDensity,
		RailActivity: stored.RailActivity,
	}
	if stored.UpdateCheck != nil {
		p.UpdateCheck = *stored.UpdateCheck
	}
	if !validView(p.View) {
		p.View = "focus"
	}
	if !validDensity(p.Density) {
		p.Density = "2x2"
	}
	if !validUsageModel(p.UsageModel) {
		p.UsageModel = defaultUsageModel
	}
	if !validRailSort(p.RailSort) {
		p.RailSort = defaultRailSort
	}
	if !validTheme(p.Theme) {
		p.Theme = defaultTheme
	}
	if !validRailDensity(p.RailDensity) {
		p.RailDensity = defaultRailDensity
	}
	if !validRailActivity(p.RailActivity) {
		p.RailActivity = defaultRailActivity
	}
	return p
}

// checkEnabledSetter is prefsFeature's narrow view of the update feature: prefs notifies
// it of a checkEnabled transition. update's own initial value comes from its own
// loadPrefs read inside newUpdateFeature, not from a *prefsFeature — so update is built
// first and handed straight into newPrefsFeature; there is no construction-order cycle to
// break.
type checkEnabledSetter interface {
	SetCheckEnabled(enabled bool)
}

// prefsFeature owns PUT /api/prefs and the snapshot's prefs object (plan code-breakup
// REQ-6).
type prefsFeature struct {
	store         *store.Store
	hub           *wsHub
	updateChecker checkEnabledSetter
	log           zerolog.Logger
}

func newPrefsFeature(st *store.Store, hub *wsHub, updateChecker checkEnabledSetter, log zerolog.Logger) *prefsFeature {
	return &prefsFeature{store: st, hub: hub, updateChecker: updateChecker, log: log}
}

func (f *prefsFeature) mount(mux *http.ServeMux, guard func(http.Handler) http.Handler) {
	mux.Handle("PUT /api/prefs", guard(http.HandlerFunc(f.handlePutPrefs)))
}

func (f *prefsFeature) contribute(ctx context.Context, snap *Snapshot) {
	snap.Prefs = loadPrefs(ctx, f.store)
}

// prefsField is one validated string-ish pref in handlePutPrefs's table. present says the
// request carried the field at all; valid and apply are only ever called when it did, so
// both may dereference the request pointer unguarded.
type prefsField struct {
	present bool
	valid   func() bool
	msg     string
	apply   func(*PrefsInfo)
}

// prefsFields is handlePutPrefs's validate-then-apply table. Slice order is load-bearing:
// it decides which single error a request that is invalid in several fields at once
// reports, so entries must stay in the order view, density, usageModel, railSort, theme,
// railDensity, railActivity. UpdateCheck is deliberately absent — it is a bool with a
// changed-tracking side effect, handled on its own in handlePutPrefs.
func prefsFields(req *prefsRequest) []prefsField {
	return []prefsField{
		{
			present: req.View != nil,
			valid:   func() bool { return validView(*req.View) },
			msg:     "view must be one of focus, tiles",
			apply:   func(p *PrefsInfo) { p.View = *req.View },
		},
		{
			present: req.Density != nil,
			valid:   func() bool { return validDensity(*req.Density) },
			msg:     "density must be one of 2x2, 3x2",
			apply:   func(p *PrefsInfo) { p.Density = *req.Density },
		},
		{
			present: req.UsageModel != nil,
			valid:   func() bool { return validUsageModel(*req.UsageModel) },
			msg:     "usageModel must be 1-32 characters after trim",
			apply:   func(p *PrefsInfo) { p.UsageModel = strings.TrimSpace(*req.UsageModel) },
		},
		{
			present: req.RailSort != nil,
			valid:   func() bool { return validRailSort(*req.RailSort) },
			msg:     "railSort must be one of manual, attention",
			apply:   func(p *PrefsInfo) { p.RailSort = *req.RailSort },
		},
		{
			present: req.Theme != nil,
			valid:   func() bool { return validTheme(*req.Theme) },
			msg:     "theme must be 1-32 chars of a-z, 0-9 or -, starting with a letter",
			apply:   func(p *PrefsInfo) { p.Theme = *req.Theme },
		},
		{
			present: req.RailDensity != nil,
			valid:   func() bool { return validRailDensity(*req.RailDensity) },
			msg:     "railDensity must be one of compact, comfortable, expanded",
			apply:   func(p *PrefsInfo) { p.RailDensity = *req.RailDensity },
		},
		{
			present: req.RailActivity != nil,
			valid:   func() bool { return validRailActivity(*req.RailActivity) },
			msg:     "railActivity must be one of turn, prompt, reply, both",
			apply:   func(p *PrefsInfo) { p.RailActivity = *req.RailActivity },
		},
	}
}

// handlePutPrefs is PUT /api/prefs: validates, persists the merged prefs object to kv,
// and broadcasts the full object to every UI socket (INV-4).
func (f *prefsFeature) handlePutPrefs(w http.ResponseWriter, r *http.Request) {
	var req prefsRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSONError(w, http.StatusBadRequest, "invalid_request", "invalid JSON body")
		return
	}

	// One pass, because a field that is absent can never be invalid: the nothing-supplied
	// error below is therefore still unreachable whenever any field-level error fires,
	// exactly as it was when the two checks were written out separately.
	fields := prefsFields(&req)
	supplied := req.UpdateCheck != nil
	for _, field := range fields {
		if !field.present {
			continue
		}
		supplied = true
		if !field.valid() {
			writeJSONError(w, http.StatusBadRequest, "invalid_request", field.msg)
			return
		}
	}
	if !supplied {
		writeJSONError(w, http.StatusBadRequest, "invalid_request", "at least one of view, density, usageModel, railSort, theme, railDensity, railActivity or updateCheck is required")
		return
	}

	ctx := r.Context()
	prefs := loadPrefs(ctx, f.store)
	updateCheckChanged := false
	for _, field := range fields {
		if field.present {
			field.apply(&prefs)
		}
	}
	if req.UpdateCheck != nil && *req.UpdateCheck != prefs.UpdateCheck {
		prefs.UpdateCheck = *req.UpdateCheck
		updateCheckChanged = true
	}

	encoded, err := json.Marshal(prefs)
	if err != nil {
		f.log.Error().Err(err).Msg("encoding prefs failed")
		writeJSONError(w, http.StatusInternalServerError, "internal_error", msgInternalError)
		return
	}
	if err := f.store.KVSet(ctx, prefsKVKey, string(encoded)); err != nil {
		f.log.Error().Err(err).Msg("persisting prefs failed")
		writeJSONError(w, http.StatusInternalServerError, "internal_error", msgInternalError)
		return
	}

	f.hub.broadcast(prefsMessage{Type: "prefs", Prefs: prefs})

	// REQ-3: the check-enabled side effect fires only on an actual transition — a PUT
	// that merely re-states the current value must not clear an in-flight check's result
	// or force an extra immediate poll.
	if updateCheckChanged && f.updateChecker != nil {
		f.updateChecker.SetCheckEnabled(prefs.UpdateCheck)
	}

	w.WriteHeader(http.StatusNoContent)
}

File: web/src/protocol/prefs.ts
// Wire types and parser for `Prefs` and its `prefs` broadcast (docs/protocol.md,
// kb:anchor/prefs.put) — one of the protocol/ concept modules split
// out of the former protocol.ts.

import { isBoolean, isRecord, isString } from "./decode";

// kb:anchor/prefs.put: the rail's sort mode pref.
export const RAIL_SORTS = ["manual", "attention"] as const;
export type RailSort = (typeof RAIL_SORTS)[number];

export const VIEWS = ["focus", "tiles"] as const;
export type View = (typeof VIEWS)[number];

export const DENSITIES = ["2x2", "3x2"] as const;
export type Density = (typeof DENSITIES)[number];

// kb:anchor/prefs.put: the rail/strip card density pref.
export const RAIL_DENSITIES = ["compact", "comfortable", "expanded"] as const;
export type RailDensity = (typeof RAIL_DENSITIES)[number];

// kb:anchor/prefs.put: which text a card's activity line
// shows — turn-aware by default (kb:adr/rail-activity-line-turn-aware-default-with-pref).
export const RAIL_ACTIVITIES = ["turn", "prompt", "reply", "both"] as const;
export type RailActivity = (typeof RAIL_ACTIVITIES)[number];

// kb:anchor/prefs.put: density is always present alongside view — the daemon's default
// before any PUT is {"view":"focus","density":"2x2","usageModel":"Fable"} — so all three
// fields are required here, matching every real `snapshot`/`prefs` payload on the wire.
// `parsePrefs` below defaults a missing `usageModel` key to `"Fable"` so a payload without
// it still parses.
export interface Prefs {
  view: View;
  density: Density;
  usageModel: string;
  // kb:anchor/prefs.put: defaulted to "manual" client-side when
  // the key is absent (same tolerance as `usageModel`, for a daemon predating this field).
  railSort: RailSort;
  // kb:anchor/prefs.put: opaque to the daemon
  // beyond its pattern — the client owns the theme registry (theme.ts). Missing key
  // (a daemon predating this field) defaults to "follow", same tolerance as usageModel/railSort.
  theme: string;
  // kb:anchor/prefs.put / kb:anchor/ws.prefs: governs checking only - the binary
  // never changes without an explicit apply. Missing key (a daemon predating this field) defaults to
  // true (the daemon's own documented default), same tolerance as the other prefs above.
  updateCheck: boolean;
  // kb:anchor/prefs.put: card density in the rail and the
  // Tiles strip. Missing key (a daemon predating this field) defaults to "comfortable", same tolerance
  // as the other prefs above.
  railDensity: RailDensity;
  // kb:anchor/prefs.put: which text a card's activity line
  // shows. Missing key (a daemon predating this field) defaults to "turn", same tolerance as above.
  railActivity: RailActivity;
}

// One-owner defaults for every pref — `parsePrefsField`'s
// fallback for a missing wire key below, and `app.ts`'s initial `AppState` before the
// first snapshot ever arrives. `view`/`density` have no *wire* default (every daemon
// always sends both, so `parsePrefs` rejects a payload missing either rather than
// defaulting it) — they're here only because the client still needs something to paint
// before it has heard from the daemon at all, and kb:anchor/ws.prefs documents the
// daemon's own default snapshot as exactly this object.
export const PREF_DEFAULTS: Prefs = {
  view: "focus",
  density: "2x2",
  usageModel: "Fable",
  railSort: "manual",
  theme: "follow",
  updateCheck: true,
  railDensity: "comfortable",
  railActivity: "turn",
};

// kb:anchor/ws.prefs: full-object echo of every accepted `PUT /api/prefs`,
// broadcast to every connected UI socket.
export interface PrefsMessage {
  type: "prefs";
  prefs: Prefs;
}

function isRailSort(value: unknown): value is RailSort {
  return (RAIL_SORTS as readonly unknown[]).includes(value);
}

function isView(value: unknown): value is View {
  return (VIEWS as readonly unknown[]).includes(value);
}

function isDensity(value: unknown): value is Density {
  return (DENSITIES as readonly unknown[]).includes(value);
}

// Exported: features/rail.ts's density-button guard imports this instead of keeping its
// own copy.
export function isRailDensity(value: unknown): value is RailDensity {
  return (RAIL_DENSITIES as readonly unknown[]).includes(value);
}

// Exported: render/settings.ts's rail-activity-radio guard imports this instead of
// keeping its own copy.
export function isRailActivity(value: unknown): value is RailActivity {
  return (RAIL_ACTIVITIES as readonly unknown[]).includes(value);
}

/** Reads one prefs field that has a fixed default for a missing key and a guard for a
 * present one — every field below `view`/`density` (which have no default; every
 * daemon always sends them) shares this exact "missing key defaults, present-but-wrong-shape
 * rejects the whole object" shape. Extracted so `parsePrefs` states each field once,
 * under Biome's complexity ceiling, rather than repeating a ternary-then-guard pair per
 * field inline. */
function parsePrefsField<T>(raw: unknown, fallback: T, guard: (v: unknown) => v is T): T | null {
  const v = raw === undefined ? fallback : raw;
  return guard(v) ? v : null;
}

export function parsePrefs(value: unknown): Prefs | null {
  if (!isRecord(value)) return null;
  const view = value["view"];
  const density = value["density"];
  if (!isView(view)) return null;
  if (!isDensity(density)) return null;
  // Missing key (a daemon predating this field) defaults to the daemon's own
  // documented default (kb:anchor/ws.prefs, PREF_DEFAULTS above).
  const usageModel = parsePrefsField(value["usageModel"], PREF_DEFAULTS.usageModel, isString);
  if (usageModel === null) return null;
  // Missing key (a daemon predating this field) defaults to PREF_DEFAULTS.railSort
  // (kb:anchor/prefs.put's documented default).
  const railSort = parsePrefsField(value["railSort"], PREF_DEFAULTS.railSort, isRailSort);
  if (railSort === null) return null;
  // Missing key (a daemon predating this field) defaults to
  // PREF_DEFAULTS.theme (kb:anchor/prefs.put's documented default) — the daemon treats
  // the string as opaque beyond its pattern, so no further validation happens client-side.
  const theme = parsePrefsField(value["theme"], PREF_DEFAULTS.theme, isString);
  if (theme === null) return null;
  // Missing key (a daemon predating this field) defaults to PREF_DEFAULTS.updateCheck
  // (kb:anchor/prefs.put's documented default), same tolerance as usageModel/railSort/theme above.
  const updateCheck = parsePrefsField(value["updateCheck"], PREF_DEFAULTS.updateCheck, isBoolean);
  if (updateCheck === null) return null;
  // Missing key (a daemon predating this field) defaults to
  // PREF_DEFAULTS.railDensity (kb:anchor/prefs.put's documented default). A *stored*
  // value outside the enum is the daemon's problem, not the wire's — it falls back to
  // "comfortable" server-side before it is ever sent. If an
  // out-of-enum railDensity somehow reaches the client anyway, parsePrefsField's guard
  // fails and parsePrefs rejects the whole prefs object (protocol/prefs.test.ts's "rejects a
  // railDensity value outside the compact|comfortable|expanded enum").
  const railDensity = parsePrefsField(
    value["railDensity"],
    PREF_DEFAULTS.railDensity,
    isRailDensity,
  );
  if (railDensity === null) return null;
  // Missing key (a daemon predating this field) defaults to
  // PREF_DEFAULTS.railActivity (kb:anchor/prefs.put's documented default), same tolerance
  // as railDensity above.
  const railActivity = parsePrefsField(
    value["railActivity"],
    PREF_DEFAULTS.railActivity,
    isRailActivity,
  );
  if (railActivity === null) return null;
  return {
    view,
    density,
    usageModel,
    railSort,
    theme,
    updateCheck,
    railDensity,
    railActivity,
  };
}

export function parsePrefsMessage(rec: Record<string, unknown>): PrefsMessage | null {
  const prefs = parsePrefs(rec["prefs"]);
  if (!prefs) return null;
  return { type: "prefs", prefs };
}
